### DISCUSSION TITLE
OpenTelemetry Trace-ID header for Responses

### DISCUSSION THREAD

**Discussion by wodka** [vercel/next.js#56071] [Help] [answered]
### Summary

It would be nice to add a header for every response if tracing is enabled with the trace-id.

I used to do this with a custom server and setting it manually for the response. With the new https://nextjs.org/docs/pages/building-your-application/optimizing/open-telemetry I could not yet find a way how to do this.

```
const requestListener = async (req: IncomingMessage, res: ServerResponse): Promise<void> => {
  const span = trace.getSpan(context.active())

  if (span) {
    const {traceId} = span.spanContext()
    res.setHeader('trace-id', traceId)
  }
```

**Comment by VadimZhiltsov** [comment]
Any updates so far? That seems to be an important part of application observability, but after so many time I also was not able to find a good way to return trace id in response

**Comment by Netail** [comment]
We have an incoming trace ID as a header, any idea how to utilize this within Nextjs?

**Comment by tageecc** [comment]
Any updates?

**Comment by wodka** [comment, marked as answer]
2 years later and I have a solution - in the proxy.ts file add the following:

```ts
import { trace } from "@opentelemetry/api";
import { NextResponse } from "next/server";

export const proxy = (req) => {
	const res = NextResponse.next();

	const traceId = trace.getActiveSpan()?.spanContext().traceId;

	if (traceId) {
		// awesome, we have a trace -> now expose it!
		res.headers.set("x-trace-id", traceId);
	}

	return res;
};
```

if there is an active span it will then append the trace id to the headers of the response.
